essence-of-live-coding/src/LiveCoding/Cell/Profunctor.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE DeriveAnyClass #-}
+
+-- export only instances
+module LiveCoding.Cell.Profunctor (module X) where
+
+-- base
+import Data.Coerce ( coerce, Coercible )
+import Control.Applicative (liftA2)
+
+-- profunctors
+import Data.Profunctor
+import Data.Profunctor
+    ( Strong, WrappedArrow(WrapArrow), Profunctor )
+import Data.Profunctor.Strong ( Strong )
+import Data.Profunctor.Choice ( Choice )
+import Data.Profunctor.Traversing as X
+import Control.Monad.Trans.State (StateT(runStateT, StateT))
+
+-- essence-of-live-coding
+import LiveCoding.Cell ( Cell(Cell, ArrM) )
+
+deriving via (WrappedArrow (Cell m)) instance Monad m => Profunctor (Cell m) 
+deriving via (WrappedArrow (Cell m)) instance Monad m => Strong (Cell m)
+deriving via (WrappedArrow (Cell m)) instance Monad m => Data.Profunctor.Choice.Choice (Cell m) 
+
+-- From https://gist.github.com/oisdk/c331222348bde9dfdcad8ca39739e421
+newtype Strict a = Strict { unStrict :: a }
+
+instance Functor Strict where
+  fmap f x = let !vx = unStrict x in Strict (f vx)
+  {-# INLINE fmap #-}
+  x <$ xs = let !_ = unStrict xs in Strict x
+  {-# INLINE (<$) #-}
+  
+instance Applicative Strict where
+  pure = Strict
+  {-# INLINE pure #-}
+  fs <*> xs = let !vx = unStrict xs in Strict (unStrict fs vx)
+  {-# INLINE (<*>) #-}
+  xs *> ys = let !_ = unStrict xs in ys
+  {-# INLINE (*>) #-}
+  xs <* ys = let !_ = unStrict ys in xs
+  {-# INLINE (<*) #-}
+  
+fmap' :: Traversable f => (a -> b) -> f a -> f b
+fmap' = (coerce :: ((a -> Strict b) -> f a -> Strict (f b)) -> (a -> b) -> f a -> f b) traverse
+{-# INLINE fmap' #-}
+
+newtype StrictT f a = StrictT { unStrictT :: f a }
+
+instance Functor f => Functor (StrictT f) where
+  fmap f = StrictT #. fmap (\ !vx -> f vx) .# unStrictT
+  {-# INLINE fmap #-}
+ 
+(#.) :: Coercible b c => (b -> c) -> (a -> b) -> a -> c
+(#.) _ = coerce
+{-# INLINE (#.) #-}
+
+(.#) :: Coercible a b => (b -> c) -> (a -> b) -> a -> c
+(.#) f _ = coerce f
+{-# INLINE (.#) #-}
+
+instance Applicative f => Applicative (StrictT f) where
+  pure = StrictT #. pure
+  {-# INLINE pure #-}
+  (<*>) = (coerce :: (f (a -> b) -> f a -> f b) -> (StrictT f (a -> b) -> StrictT f a -> StrictT f b)) (liftA2 (\fs !vx -> fs vx))
+  {-# INLINE (<*>) #-}
+  liftA2 f = (coerce :: (f a -> f b -> f c) -> (StrictT f a -> StrictT f b -> StrictT f c)) (liftA2 (\ !x !y -> f x y))
+  {-# INLINE liftA2 #-}
+
+traverseStrict :: (Traversable t, Applicative f) => (a -> f b) -> t a -> f (t b)
+traverseStrict = (coerce :: ((a -> StrictT f b) -> t a -> StrictT f (t b)) -> (a -> f b) -> t a -> f (t b)) traverse
+{-# INLINE traverseStrict #-}
+
+instance Monad m => Traversing (Cell m) where
+  traverse' (Cell state step) = Cell state step' where
+    step' s a = runStateT (traverse (\a -> StateT (`step` a)) a) s
+  traverse' (ArrM f) = ArrM (traverseStrict f)
```

Comment:
The `StrictT` trick is very interesting, but I have a bad feeling to introduce so much generic code which should reside in a separate package, just for this one special case.

I don't understand why `traverseStrict` is needed here, and not in the `Cell` case. And maybe it's possible to implement `traverseStrict` in a simpler way?

Also, I don't understand what exactly is strict here. The container structure we're traversing over? Or the individual values? Because the latter doesn't make sense to me (even though I implemented `resampleList` that way).

Reply:
> I don't understand why `traverseStrict` is needed here, and not in the `Cell` case. And maybe it's possible to implement `traverseStrict` in a simpler way?

Ah, you said that it's because of strict state! Makes sense.